pkg/deploy/strategy/rolling/rolling.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	// HACK: Assign the source ID annotation that the rolling updater expects,
+	// unless it already exists on the deployment.
+	if _, hasSourceId := deployment.Annotations[sourceIdAnnotation]; !hasSourceId {
+		deployment.Annotations[sourceIdAnnotation] = fmt.Sprintf("%s:%s", latest.Name, latest.ObjectMeta.UID)
+		if updated, err := s.client.UpdateReplicationController(deployment.Namespace, deployment); err != nil {
+			return fmt.Errorf("couldn't assign source annotation to deployment %s/%s: %v", deployment.Namespace, deployment.Name, err)
+		} else {
+			deployment = updated
+		}
+	}
+
+	// HACK: There's a validation in the rolling updater which assumes that when
+	// an existing RC is supplied, it will have >0 replicas- a validation which
+	// is then disregarded as the desired count is obtained from the annotation
+	// on the RC. For now, fake it out by just setting replicas to 1.
+	deployment.Spec.Replicas = 1
+
+	// Perform a rolling update.
+	params := config.Template.Strategy.RollingParams
+	rollingConfig := &kubectl.RollingUpdaterConfig{
+		Out:           &rollingUpdaterWriter{},
+		OldRc:         latest,
+		NewRc:         deployment,
+		UpdatePeriod:  time.Duration(*params.UpdatePeriodSeconds) * time.Second,
+		Interval:      time.Duration(*params.IntervalSeconds) * time.Second,
+		Timeout:       time.Duration(*params.TimeoutSeconds) * time.Second,
+		CleanupPolicy: kubectl.PreserveRollingUpdateCleanupPolicy,
+	}
+	return s.rollingUpdate(rollingConfig)
+}
+
+// findLatestDeployment retrieves deployments identified by oldDeployments and
+// returns the latest one from the list, or nil if there are no old
+// deployments.
+func (s *RollingDeploymentStrategy) findLatestDeployment(oldDeployments []kapi.ObjectReference) (*kapi.ReplicationController, error) {
+	// Find the latest deployment from the list of old deployments.
+	var latest *kapi.ReplicationController
+	latestVersion := 0
+	for _, obj := range oldDeployments {
+		deployment, err := s.client.GetReplicationController(obj.Namespace, obj.Name)
+		if err != nil {
+			if !kerrors.IsNotFound(err) {
+				return nil, fmt.Errorf("couldn't get old deployment %s/%s: %v", obj.Namespace, obj.Name)
+			}
+			glog.Infof("Ignoring missing old deployment %s/%s", obj.Namespace, obj.Name)
+			continue
+		}
+
+		if val, hasVersion := deployment.Annotations[deployapi.DeploymentVersionAnnotation]; hasVersion {
+			version, err := strconv.Atoi(val)
+			if err != nil {
+				return nil, fmt.Errorf("deployment %s/%s has invalid version annotation value '%s': %v", deployment.Namespace, deployment.Name, val, err)
+			}
+			if version > latestVersion {
+				latest = deployment
+				latestVersion = version
+			}
+		} else {
+			glog.Infof("Ignoring deployment with missing version annotation: %s/%s", deployment.Namespace, deployment.Name)
+		}
+	}
+	return latest, nil
+}
+
+type rollingUpdaterClient struct {
+	GetReplicationControllerFn     func(namespace, name string) (*kapi.ReplicationController, error)
```

Comment:
----- Original Message -----
> > +			if err != nil {
> > +				return nil, fmt.Errorf("deployment %s/%s has invalid version
> > annotation value '%s': %v", deployment.Namespace, deployment.Name, val,
> > err)
> > +			}
> > +			if version > latestVersion {
> > +				latest = deployment
> > +				latestVersion = version
> > +			}
> > +		} else {
> > +			glog.Infof("Ignoring deployment with missing version annotation:
> > %s/%s", deployment.Namespace, deployment.Name)
> > +		}
> > +	}
> > +	return latest, nil
> > +}
> > +
> > +type rollingUpdaterClient struct {
> > +	GetReplicationControllerFn     func(namespace, name string)
> > (*kapi.ReplicationController, error)
> 
> I'm not seeing any pluggable ReplicationControllerNamespacer upstream I can
> use.

Now is your chance.  I'd rather you wrote the generic one here rather than invent YACMI

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2127/files#r29960924